// naoko-player/lib/core/MusicManager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const voice_1 = require("@discordjs/voice");
const MusicNode_1 = __importDefault(require("./MusicNode"));
class MusicManager {
    constructor(options) {
        this.options = options;
        this.nodes = new Map();
        this.providers = new Map();
        for (const perv of options.providers) {
            this.providers.set(perv.provider, perv);
        }
    }
    createNode(guildId, force) {
        if (!force && this.nodes.has(guildId))
            return this.nodes.get(guildId);
        const node = new MusicNode_1.default(guildId, this);
        if (!this.nodes.has(guildId))
            this.nodes.set(guildId, node);
        return node;
    }
    getNode(guildId, forceCreate) {
        if (forceCreate && !this.nodes.has(guildId))
            return this.createNode(guildId);
        return this.nodes.get(guildId);
    }
    /**
     * Creates a VoiceConnection to a Discord voice channel.
     *
     * @param options the options for joining the voice channel
     */
    get joinVoiceChannel() {
        return voice_1.joinVoiceChannel;
    }
}
module.exports = MusicManager;
